google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -2121,6 +2129,8 @@ private void handleCreateSessionsFailure(SpannerException e, int count) {
       }
       if (isDatabaseOrInstanceNotFound(e)) {
         setResourceNotFoundException((ResourceNotFoundException) e);
+        pendingClosure += 1;
```

Comment:
This is AFAICT redundant. It will be overwritten on line 2170 when the pool is closed.

Reply:
The issue is that if I don't increment the `pendingClosure`, when the `PoolMaintainer.close()` executes and `running` is marked as false, it won't go into the branch on line 2145 (`closureFuture.set(null)`), since the `pendingClosure` will be `-1`. I did not want to change the condition to `pendingClosure <= 0` instead of `pendingClosure == 0` in line 2144, I preferred to increment the closures here.

Reply:
I've been trying to go through it once more, and I actually think that it is not only redundant, but also possibly a source of a `NullPointerException`. What's important here is that `closureFuture` is only defined and only relevant once the `SessionPool` is closing. It is only instantiated during the `closeAsync()` method [here](https://github.com/googleapis/java-spanner/blob/6cec2918d33d61a19772b40ea665618228a20243/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java#L2167).

So what might happen in this case is:
1. A `DatabaseNotFoundException` is thrown during the creation of new sessions.
2. The `handleCreateSessionsFailure` method sets the `pendingClosure` is to 1.
2. `poolMaintainer.close()` is called. If `running` is false, `pendingClosure` is decreased and `closureFuture.set(null)` is called. However, `closureFuture` will be `null` at this moment, unless the `SessionPool` has already been closed.

The reason that I thought it was redundant is that we are initializing the value of `pendingClosure` one [line 2170](https://github.com/googleapis/java-spanner/blob/6cec2918d33d61a19772b40ea665618228a20243/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java#L2170) and then increasing the `pendingClosure` value on [line 2173](https://github.com/googleapis/java-spanner/blob/6cec2918d33d61a19772b40ea665618228a20243/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java#L2173) if the `PoolMaintainer` is still running. So this part will ensure that `pendingClosure` will have the correct value and will reach zero once everything is closed.

Or am I missing something else here?

Reply:
You are correct, an NPE is thrown in that case (verified with `ReadAsyncTest#invalidDatabase`). Removed the `pendingClosure += 1` from the code.

The weird thing is that the exception there was being swollen, because it was part of `SessionClient.BatchCreateSessionsRunnable#run` method, so I never got to see it.